api/src/server/auth.rs

```diff
@@ -199,50 +152,63 @@ pub async fn route_login(
-            .optional()
-            .map_err(ResponseError::from)?;
-    match existing_user_provider {
-        Some(user_provider) => match user_provider.user_id {
-            Some(_) => {
-                // User already exists so normal login
-                Ok(log_in_user(&user_provider, &identity, redirect_dest))
-            }
-            None => {
-                // no user account associated with this user_provider
-                // yet so make one (they have logged in but did not set
-                // username)
-                init_user(&user_provider, userinfo, conn)?;
-                Ok(log_in_user(&user_provider, &identity, redirect_dest))
-            }
-        },
-        None => {
-            // user_provider not found (first login) so make the row
-            // then init the user
-            let user_provider: UserProvider = NewUserProvider {
+
+    // Insert the sub+provider, or just return the existing one if it's already
+    // in the DB. We need to do this in a transaction to prevent race conditions
+    // if the provider gets deleted in another thread.
+    let user_provider_id: Uuid =
+        conn.transaction::<Uuid, ResponseError, _>(|| {
+            // Insert, if the row already exists, just return None
+            let inserted = NewUserProvider {
                 sub,
                 provider_name,
                 user_id: None,
             }
             .insert()
-            .returning(user_providers::all_columns)
+            .on_conflict_do_nothing()
+            .returning(user_providers::columns::id)
             .get_result(conn)
-            .unwrap();
+            .optional()
+            .map_err(ResponseError::from)?;
 
-            // Create a new User object, then set the auth cookie
-            init_user(&user_provider, userinfo, conn)?;
-            Ok(log_in_user(&user_provider, &identity, redirect_dest))
-        }
-    }
+            match inserted {
+                // Insert didn't return anything, which means the row is already
+                // in the DB. Just select that row.
+                None => user_providers::table
+                    .select(user_providers::columns::id)
+                    .filter(user_providers::columns::sub.eq(sub))
+                    .filter(
+                        user_providers::columns::provider_name
+                            .eq(provider_name),
+                    )
+                    .get_result(conn)
+                    .map_err(ResponseError::from),
+                Some(inserted_id) => Ok(inserted_id),
+            }
+        })?;
+
+    // Add a cookie which can be used to auth requests. We use the UserProvider
+    // ID so that this works even if the User object hasn't been created yet.
+    identity.remember(user_provider_id.to_string());
+
+    // Redirect to the path specified in the OpenID state param
+    Ok(HttpResponse::Found()
+        .header(http::header::LOCATION, redirect_dest)
+        .finish())
```

Comment:
This is mostly just cleanup. I deleted the `init_user` function since we don't need it anymore, and rearranged the logic here to try to clean it up. I also got rid of the `log_in_user` function since it was only a couple lines and after moving the logic around, it was only being called once from here anyway